modules/packages/Collection.chpl

```diff
@@ -18,118 +18,212 @@
  */
 
 /*
-  A 'Collection' is a data structure, a container for elements that provide support
-  for insert, lookup, remove, and iteration operations.
+  
+  Summary
+  _______
+
+  A 'Collection' is an interface for a data structure that is a 'contract' between
+  the user and implementer that it fulfills the following:
+
+  1. Is safe parallel-safe, hence is safe to use across multiple tasks across multiple locales.
+  2. Supports the basic operations that any data structure needs to be truly useful, that is:
+
+    a. Insertion of an arbitrary element. From this, we can insert bulk arbitrary elements.
+    b. Removal of an arbitrary element. From this, we can remove bulk arbitrary elements.
+    c. Iteration over all elements. From this, we can perform lookups over all elements.
+  
+  From the standpoint of the user, who directly benefits, they obtain a very nice but minimal
+  guarantee on the object they are using, and from the implementor's view, they get to the
+  benefit from both implementing a well-design interface, and from getting some utility methods 
+  'for-free'.
+
+  .. note::
+
+    The documentation for the Collection modules are being incrementally revised and improved.
 */
-class Collection {
-  /*
-    The type of element that this Collection holds.
-  */
-  type eltType;
 
-  /*
-    Adds an element to this data structure.
-  */
-  inline proc add(elt : eltType) : bool {
-    halt("'proc add(elt : eltType) : bool' is not supported...");
-  }
+module Collection {
+  class CollectionImpl {
+    /*
+      The type of element that this Collection holds.
+    */
+    type eltType;
+
+    /*
+      Adds an element to this data structure.
+    */
+    proc add(elt : eltType) : bool {
+      halt("'proc add(elt : eltType) : bool' is not supported...");
+    }
+
+    /*
+      Add all elements in bulk to this data structure. If the data structure
+      rejects an element, we cease to offer more. We return the number of elements
+      successfully added to this data structure.
+
+      **Warning:** While this method will add as many items as possible, it will `break`
+      when it is unable to consume more elements. Due to the fact that using `break` in
+      an iterator will cause a memory leak and potentially leave this data structure in
+      a undefined state, this should not be used to append another data structure to a
+      bounded structure unless it is known it will succeed.
+    */
+    proc addBulk(elts) : int {

```

Comment:
I think I prefer `bulkAdd` (similarly `bulkRemove`). That's what sparse domains use. Sparse domains and collections are not closely related, but nonetheless..

Reply:
Will do.